File: 3.x.x/crates-utopia/auth/src/hashes/phpass.rs
//! `PHPass` password hashing (legacy).

use std::collections::HashMap;
use std::time::{SystemTime, UNIX_EPOCH};

use md5::{Digest, Md5};
use rand::RngCore;
use serde_json::{json, Value};

use crate::error::AuthError;
use crate::hash::{Hash, HashMut, HashOptions};

const ITOA64: &[u8; 64] = b"./0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";

/// `PHPass` hasher with support for the portable `$P$`/`$H$` MD5 algorithm.
#[derive(Debug, Clone)]
pub struct PHPass {
    inner: HashOptions,
}

impl Default for PHPass {
    fn default() -> Self {
        Self::new()
    }
}

impl PHPass {
    /// Create a `PHPass` hasher with PHP-compatible defaults.
    #[must_use]
    pub fn new() -> Self {
        let random_state = SystemTime::now().duration_since(UNIX_EPOCH).map_or_else(
            |_| String::new(),
            |duration| duration.as_nanos().to_string(),
        );

        let mut inner = HashOptions::new();
        inner.options_mut().insert("type".into(), json!("phpass"));
        inner
            .options_mut()
            .insert("iteration_count_log2".into(), json!(8));
        inner
            .options_mut()
            .insert("portable_hashes".into(), json!(false));
        inner
            .options_mut()
            .insert("random_state".into(), json!(random_state));
        Self { inner }
    }

    /// Set iteration count log2 between 4 and 31.
    pub fn set_iteration_count(&mut self, count: u32) -> Result<&mut Self, AuthError> {
        if !(4..=31).contains(&count) {
            return Err(AuthError::InvalidInput(
                "Iteration count must be between 4 and 31".into(),
            ));
        }

        self.inner
            .options_mut()
            .insert("iteration_count_log2".into(), json!(count));
        Ok(self)
    }

    /// Set whether generated hashes must use the portable `PHPass` algorithm.
    pub fn set_portable_hashes(&mut self, portable: bool) -> &mut Self {
        self.inner
            .options_mut()
            .insert("portable_hashes".into(), json!(portable));
        self
    }

    fn portable_hashes(&self) -> bool {
        self.inner
            .get_option("portable_hashes")
            .and_then(Value::as_bool)
            .unwrap_or(false)
    }

    fn iteration_count_log2(&self) -> Result<u32, AuthError> {
        self.inner.require_u32("iteration_count_log2")
    }

    fn get_random_bytes(count: usize) -> Result<Vec<u8>, AuthError> {
        if count < 1 {
            return Err(AuthError::InvalidInput(
                "Argument count must be a positive integer".into(),
            ));
        }

        let mut output = vec![0u8; count];
        rand::thread_rng().fill_bytes(&mut output);
        Ok(output)
    }

    fn encode64(input: &[u8], count: usize) -> Result<String, AuthError> {
        if count < 1 {
            return Err(AuthError::InvalidInput(
                "Argument count must be a positive integer".into(),
            ));
        }

        let mut output = String::new();
        let mut i = 0;
        while i < count {
            let mut value = u32::from(input[i]);
            i += 1;
            output.push(ITOA64[(value & 0x3f) as usize] as char);

            if i < count {
                value |= u32::from(input[i]) << 8;
            }
            output.push(ITOA64[((value >> 6) & 0x3f) as usize] as char);

            if i >= count {
                break;
            }
            i += 1;

            if i < count {
                value |= u32::from(input[i]) << 16;
            }
            output.push(ITOA64[((value >> 12) & 0x3f) as usize] as char);

            if i >= count {
                break;
            }
            i += 1;

            output.push(ITOA64[((value >> 18) & 0x3f) as usize] as char);
        }

        Ok(output)
    }

    fn gensalt_private(&self, input: &[u8]) -> Result<String, AuthError> {
        let count = (self.iteration_count_log2()? + 5).min(30);
        Ok(format!(
            "$P${}{}",
            ITOA64[count as usize] as char,
            Self::encode64(input, 6)?
        ))
    }

    fn crypt_private(&self, password: &str, setting: &str) -> Result<String, AuthError> {
        let mut output = "*0".to_owned();
        if setting.starts_with(&output) {
            output.clear();
            output.push_str("*1");
        }

        if !(setting.starts_with("$P$") || setting.starts_with("$H$")) || setting.len() < 12 {
            return Ok(output);
        }

        let setting_bytes = setting.as_bytes();
        let Some(count_log2) = ITOA64.iter().position(|value| *value == setting_bytes[3]) else {
            return Ok(output);
        };
        if !(7..=30).contains(&count_log2) {
            return Ok(output);
        }

        let count = 1u64 << count_log2;
        let salt = &setting_bytes[4..12];
        let mut input = Vec::with_capacity(salt.len() + password.len());
        input.extend_from_slice(salt);
        input.extend_from_slice(password.as_bytes());
        let mut hash = Md5::digest(&input).to_vec();

        for _ in 0..count {
            let mut input = Vec::with_capacity(hash.len() + password.len());
            input.extend_from_slice(&hash);
            input.extend_from_slice(password.as_bytes());
            hash = Md5::digest(&input).to_vec();
        }

        Ok(format!("{}{}", &setting[..12], Self::encode64(&hash, 16)?))
    }
}

impl Hash for PHPass {
    fn hash(&self, value: &str) -> Result<String, AuthError> {
        #[cfg(feature = "bcrypt")]
        if !self.portable_hashes() {
            let cost = self.iteration_count_log2()?;
            if let Ok(hash) = bcrypt::hash(value, cost) {
                if hash.len() == 60 {
                    return Ok(hash);
                }
            }
        }

        let random = Self::get_random_bytes(6)?;
        let hash = self.crypt_private(value, &self.gensalt_private(&random)?)?;
        if hash.len() == 34 {
            return Ok(hash);
        }

        Ok("*".into())
    }

    fn verify(&self, value: &str, hash: &str) -> bool {
        let verification_hash = self
            .crypt_private(value, hash)
            .unwrap_or_else(|_| "*".into());
        let verification_hash = if verification_hash.starts_with('*') {
            #[cfg(feature = "bcrypt")]
            {
                return bcrypt::verify(value, hash).unwrap_or(false);
            }
            #[cfg(not(feature = "bcrypt"))]
            {
                verification_hash
            }
        } else {
            verification_hash
        };

        subtle::ConstantTimeEq::ct_eq(verification_hash.as_bytes(), hash.as_bytes()).into()
    }

    fn name(&self) -> &'static str {
        "phpass"
    }

    fn options(&self) -> &HashMap<String, Value> {
        self.inner.options()
    }
}

impl HashMut for PHPass {
    fn options_mut(&mut self) -> &mut HashMap<String, Value> {
        self.inner.options_mut()
    }
}

File: packages/auth/src/Auth/Hashes/PHPass.php
<?php

declare(strict_types=1);

namespace Utopia\Auth\Hashes;

use Utopia\Auth\Hash;

class PHPass extends Hash
{
    /**
     * Alphabet used in itoa64 conversions.
     */
    protected string $itoa64 = './0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz';

    /**
     * Constructor
     */
    public function __construct()
    {
        $randomState = microtime();
        if (\function_exists('getmypid')) {
            $randomState .= getmypid();
        }

        $this->setOption('type', $this->getName());
        $this->setOption('iteration_count_log2', 8);
        $this->setOption('portable_hashes', false);
        $this->setOption('random_state', $randomState);
    }

    /**
     * {@inheritdoc}
     */
    public function hash(string $value): string
    {
        $options = $this->getOptions();
        $random = '';

        if (! $options['portable_hashes']) {
            $random = $this->getRandomBytes(16);
            $hash = crypt($value, $this->gensaltBlowfish($random));
            if (\strlen($hash) === 60) {
                return $hash;
            }
        }

        if (\strlen($random) < 6) {
            $random = $this->getRandomBytes(6);
        }

        $hash = $this->cryptPrivate($value, $this->gensaltPrivate($random));
        if (\strlen($hash) === 34) {
            return $hash;
        }

        return '*';
    }

    /**
     * {@inheritdoc}
     */
    public function verify(string $value, string $hash): bool
    {
        $verificationHash = $this->cryptPrivate($value, $hash);
        if ($verificationHash[0] === '*') {
            $verificationHash = crypt($value, $hash);
        }

        return hash_equals($hash, $verificationHash);
    }

    /**
     * Get random bytes
     */
    protected function getRandomBytes(int $count): string
    {
        if ($count < 1) {
            throw new \Exception('Argument count must be a positive integer');
        }

        $output = '';
        if (@is_readable('/dev/urandom') && ($fh = @fopen('/dev/urandom', 'rb'))) {
            $readOutput = fread($fh, $count);
            if ($readOutput !== false) {
                $output = $readOutput;
            }
            fclose($fh);
        }

        if (\strlen($output) < $count) {
            $output = '';
            $options = $this->getOptions();

            for ($i = 0; $i < $count; $i += 16) {
                $options['random_state'] = md5(microtime() . $options['random_state']);
                $output .= md5($options['random_state'], true);
            }

            $output = substr($output, 0, $count);
        }

        return $output;
    }

    /**
     * Encode in base64
     */
    protected function encode64(string $input, int $count): string
    {
        if ($count < 1) {
            throw new \Exception('Argument count must be a positive integer');
        }

        $output = '';
        $i = 0;
        do {
            $value = \ord($input[$i++]);
            $output .= $this->itoa64[$value & 0x3F];
            if ($i < $count) {
                $value |= \ord($input[$i]) << 8;
            }
            $output .= $this->itoa64[($value >> 6) & 0x3F];
            if ($i++ >= $count) {
                break;
            }
            if ($i < $count) {
                $value |= \ord($input[$i]) << 16;
            }
            $output .= $this->itoa64[($value >> 12) & 0x3F];
            if ($i++ >= $count) {
                break;
            }
            $output .= $this->itoa64[($value >> 18) & 0x3F];
        } while ($i < $count);

        return $output;
    }

    /**
     * Generate salt for private key
     */
    private function gensaltPrivate(string $input): string
    {
        $options = $this->getOptions();
        $output = '$P$';
        $output .= $this->itoa64[min($options['iteration_count_log2'] + ((PHP_VERSION >= '5') ? 5 : 3), 30)];

        return $output . $this->encode64($input, 6);
    }

    /**
     * Generate salt for Blowfish
     */
    private function gensaltBlowfish(string $input): string
    {
        $options = $this->getOptions();
        $itoa64 = './ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';

        $output = '$2a$';
        $output .= \chr(\ord('0') + \intval($options['iteration_count_log2'] / 10));
        $output .= \chr(\ord('0') + $options['iteration_count_log2'] % 10);
        $output .= '$';

        $i = 0;
        do {
            $c1 = \ord($input[$i++]);
            $output .= $itoa64[$c1 >> 2];
            $c1 = ($c1 & 0x03) << 4;
            if ($i >= 16) {
                $output .= $itoa64[$c1];
                break;
            }

            $c2 = \ord($input[$i++]);
            $c1 |= $c2 >> 4;
            $output .= $itoa64[$c1];
            $c1 = ($c2 & 0x0F) << 2;

            $c2 = \ord($input[$i++]);
            $c1 |= $c2 >> 6;
            $output .= $itoa64[$c1];
            $output .= $itoa64[$c2 & 0x3F];
        } while (1);

        return $output;
    }

    /**
     * Crypt private
     */
    private function cryptPrivate(string $password, string $setting): string
    {
        $output = '*0';
        if (substr($setting, 0, 2) === $output) {
            $output = '*1';
        }

        $id = substr($setting, 0, 3);
        // We use "$P$", phpBB3 uses "$H$" for the same thing
        if ($id !== '$P$' && $id !== '$H$') {
            return $output;
        }

        $count_log2 = strpos($this->itoa64, $setting[3]);
        if ($count_log2 < 7 || $count_log2 > 30) {
            return $output;
        }

        $count = 1 << $count_log2;
        $salt = substr($setting, 4, 8);
        if (\strlen($salt) !== 8) {
            return $output;
        }

        $hash = md5($salt . $password, true);
        do {
            $hash = md5($hash . $password, true);
        } while (--$count);

        $output = substr($setting, 0, 12);

        return $output . $this->encode64($hash, 16);
    }

    /**
     * Set iteration count (log2)
     *
     * @param  int  $count  Iteration count (log2) between 4 and 31
     *
     * @throws \InvalidArgumentException
     */
    public function setIterationCount(int $count): PHPass
    {
        if ($count < 4 || $count > 31) {
            throw new \InvalidArgumentException('Iteration count must be between 4 and 31');
        }

        $this->setOption('iteration_count_log2', $count);

        return $this;
    }

    /**
     * Set portable hashes mode
     *
     * @param  bool  $portable  Whether to use portable hashes
     */
    public function setPortableHashes(bool $portable): PHPass
    {
        $this->setOption('portable_hashes', $portable);

        return $this;
    }

    /**
     * {@inheritdoc}
     */
    public function getName(): string
    {
        return 'phpass';
    }
}
